src/core/utils/number/numberUtils.ts

```diff
@@ -0,0 +1,86 @@
+import {
+  IS_LAST_CHARACTER_DECIMAL_POINT_REGEX,
+  MATCH_ZEROS_AFTER_DECIMAL_REGEX,
+  DECIMAL_NUMBER_SEPARATOR,
+  THOUSANDTHS_SEPARATOR
+} from "./numberConstants";
+
+/**
+ * Coerces a number into a string.
+ * @param {number} value - A number to convert to string
+ * @param {number} maxFractionDigits - The decimal part length
+ * @param {string} locale - Default locale used is "en"
+ * @returns {string} The value after coercing the given value to a string.
+ */
+function numberToString(
+  value: number | string,
+  maxFractionDigits: number,
+  locale = "en"
+) {
+  let final = "";
+
+  if (typeof value === "number" && !Object.is(value, NaN)) {
+    final = value.toLocaleString(locale);
+  } else if (typeof value === "string") {
+    if (!value.includes(DECIMAL_NUMBER_SEPARATOR)) {
+      final = parseFloat(value).toLocaleString(locale);
+    }
+
+    // If the value is `"[0-9]."` or `[0-9].0+"
+    else if (
+      IS_LAST_CHARACTER_DECIMAL_POINT_REGEX.test(value) ||
+      MATCH_ZEROS_AFTER_DECIMAL_REGEX.test(value) ||
+      value.includes(DECIMAL_NUMBER_SEPARATOR)
+    ) {
+      const decimalNumberParts = value.split(DECIMAL_NUMBER_SEPARATOR);
+      const decimalPart = decimalNumberParts[1];
+      const integerPart = parseFloat(decimalNumberParts[0]).toLocaleString(locale);
+
```

Comment:
I thought we would be using Intl.NumberFormat directly instead of doing the calculations below. Rename the function to `formatNumber` and the implementation may look like this:

```
function formatNumber(
  providedOptions: Omit<Intl.NumberFormatOptions, "style"> & {
    style?: Intl.NumberFormatPartTypes | "percent";
    locale?: string;
  } = {}
) {
  const {locale, ...otherOptions} = providedOptions;
  let options = otherOptions;
  const isCurrencyFormatting = options.style === "currency";

  if (isCurrencyFormatting) {
    options = {
      currency: "USD",
      maximumFractionDigits: 5,
      minimumFractionDigits: 2,
      // "narrowSymbol" prevents showing explicit currency symbols, such as "US$" or "CA$"
      currencyDisplay: "narrowSymbol",
      ...otherOptions
    };
  } else {
    options = {
      maximumFractionDigits: 5,
      style: "decimal",
      ...otherOptions
    };
  }

  let numberFormatter: {
    format: (x: number | bigint) => string;
  };

  try {
    numberFormatter = new Intl.NumberFormat(
      locale || [navigator.language, "en-GB"],
      options
    );
  } catch (error) {
    numberFormatter = {
      format(x: number | bigint) {
        return x.toLocaleString(locale);
      }
    };

    if (isCurrencyFormatting) {
      // currencyDisplay: "narrowSymbol" option is not supported by some browsers, try by providing "symbol"
      try {
        options.currencyDisplay = "symbol";
        numberFormatter = new Intl.NumberFormat(
          locale || [navigator.language, "en-GB"],
          options
        );
      } catch (currencyFormattingError) {
        console.error({currencyFormattingError});
      }
    }
  }

  return (value: number) => {
    let formattedValue = "";

    if (!Object.is(value, NaN)) {
      formattedValue = numberFormatter.format(value);
    }

    if (formattedValue && options.currency === "USD") {
      // in case `narrowSymbol` option was failed, make sure "US$" sign appears as just "$"
      formattedValue = formattedValue.replace("US$", "$");
    }

    return formattedValue;
  };
}
```

Reply:
Thank you for your help! 🙏🏻